docs/benchmarks-design.md

```diff
@@ -0,0 +1,324 @@
+
+### Overview
+
+Superbenchark is a benchmark framework that contains lots of useful micro-benchmarks and E2E-benchmarks targeting for hardware/software validation,  performance measurement and offering advice for better resource utilization. For now, Superbenchmark suppert:
+
+Micro-benchmark: FIO, P2P_Bandwidth, TensorCore_Performance, GPU_FLOPS, TCP_Performance, RDMA_Performance, MPI_Performance, NCCL_Performance, Stress_Test.
+
+E2E-benchmark: 32-CNN models, LSTM, BERT, Seq2Seq, DeepSpeech2, NASNet, GraphSage
+
+However the current structure of all the benchmarks are very loose. The pain points we have met are as follows:
+
+1.	Every benchmark has an independent python file and there exist a lot of redundant/dead code. 
+2.	To launch one benchmark, we must specify the file path of the benchmark which is hard coded and pieced together with the test name string from config file. 
+3.	To add new micro-benchmarks for new accelerators, we must modify the related source code and add IF statement to distinguish different platforms.
+4.	To modify some common logic of E2E-benchmarks like Dataset, DataLoader and distributed related code, we need change all the E2E benchmark files.
+5.	It’s hard to extended new features, such as running benchmark for N times.
+6.	The output of each benchmark is disorganized and unstructured.
+7.	Some benchmarks have big variance, it’s hard to do the confirmation and investigation manually.
+8.	Hard to detect the interruption of running, especially for microbenchmarks which is launched by system call.
+
+This document will present a new design for the benchmarks to avoid the above pain points. 
+
+### Goals
+
+The new design have the following goals to achieve:
+
+**High Code Quality**
+•	Extract the common code into the base class, reduce the workload to maintain all benchmarks.
+•	Provide a unified entrance to avoid specifying benchmarks in hard-code way.
+**Good Extensibility**
+•	Avoid modifying existing code when add new benchmarks by using registration mechanism.
+•	Support benchmark-specific configuration, further support models with (same structure and different configs).
+**Good Usability**
+•	Unify the output format for all the micro-benchmarks and E2E-benchmarks, contains status (success or False), raw-output, error message and structured result.
+•	Can specify the run times for user to check the variance of one benchmark.
+
+## High-level Technical Design
+
+The structure of Benchmarks package can be divided into three layers from the bottom up:
+1.	Abstract base classes for all kind of benchmarks, including ModelBase, MicrobenchmarkBase and DockerBase.
+a)	ModelBase is the base class for all the public E2E models. It defines the abstract interfaces that need to be implemented by the subclasses. It has three subclasses, including PytorchModel, TFModel and ONNXModel. Each subclass will realize part of the abstract interfaces that is common for all models, such as “create_optimizer()”, “init_dataloader()”.
```

Comment:
Shall we change all to the real implementation?

I think ModelBase should be `ModelBenchmark`